### DISCUSSION TITLE
Support for union types?

### DISCUSSION THREAD

**Discussion by 3xau1o** [flix/flix#4866] [General]
Are there support for union types, like Scala, F# and Typescript?

```scala
// the following constant can only store ints from 1 to 3
val three: 1 | 2 | 3 = 3

val one: 1 = 1                     // val declaration
def foo(x: 1): Option[1] = Some(x) // param type, type arg
def bar[T <: 1](t: T): T = t       // type parameter bound
foo(1: 1)                          // type ascription
```

**Comment by magnus-madsen** [comment]
Hello @mindlid and welcome to our little community, and thanks for the question :)

We currently do not have support for union types, just plain old algebraic data types.

In the future, we are looking into a completely new approach to algebraic data types / union types, but I cannot say much more since its research and still on the drawing board. Ill update this ticket once there is more information.

**Reply by 3xau1o** [reply]
can't wait!

**Comment by magnus-madsen** [comment]
You would perhaps be interested to know that we do support extensible records: https://doc.flix.dev/records.html

**Reply by 3xau1o** [reply]
just amazing, feels more like F#/Typescript than scala/java

**Comment by magnus-madsen** [comment]
Thanks for the kind words. Let us know if you have more questions.
